packages/vue-renderer/src/renderer.js

```diff
@@ -281,190 +188,63 @@ export default class VueRenderer {
-    const renderer = context.modern ? this.renderer.modern : this.renderer.ssr
-
-    // Call ssr:context hook to extend context from modules
-    await this.context.nuxt.callHook('vue-renderer:ssr:prepareContext', context)
-
-    // Call Vue renderer renderToString
-    let APP = await renderer.renderToString(context)
-
-    // Call ssr:context hook
-    await this.context.nuxt.callHook('vue-renderer:ssr:context', context)
-    // TODO: Remove in next major release
-    await this.context.nuxt.callHook('render:routeContext', context.nuxt)
-
-    // Fallback to empty response
-    if (!context.nuxt.serverRendered) {
-      APP = `<div id="${this.context.globals.id}"></div>`
+  createRenderer() {
+    // Resource clientManifest is always required
+    if (!this.context.resources.clientManifest) {
+      return
     }
 
-    // Inject head meta
-    const m = context.meta.inject()
-    let HEAD =
-      m.title.text() +
-      m.meta.text() +
-      m.link.text() +
-      m.style.text() +
-      m.script.text() +
-      m.noscript.text()
-
-    // Add <base href=""> meta if router base specified
-    if (this.context.options._routerBaseSpecified) {
-      HEAD += `<base href="${this.context.options.router.base}">`
-    }
+    this.detectModernBuild()
 
-    // Inject resource hints
-    if (this.context.options.render.resourceHints) {
-      HEAD += this.renderSsrResourceHints(context)
+    // Create SPA renderer
+    if (this.context.resources.spaTemplate) {
+      this.renderer.spa = new SPARenderer(this.context)
     }
 
-    // Inject styles
-    HEAD += context.renderStyles()
-
-    // Serialize state
-    const serializedSession = `window.${this.context.globals.context}=${devalue(context.nuxt)};`
-    APP += `<script>${serializedSession}</script>`
-
-    // Calculate CSP hashes
-    const { csp } = this.context.options.render
-    const cspScriptSrcHashes = []
-    if (csp) {
-      // Only add the hash if 'unsafe-inline' rule isn't present to avoid conflicts (#5387)
-      const containsUnsafeInlineScriptSrc = csp.policies && csp.policies['script-src'] && csp.policies['script-src'].includes(`'unsafe-inline'`)
-      if (!containsUnsafeInlineScriptSrc) {
-        const hash = crypto.createHash(csp.hashAlgorithm)
-        hash.update(serializedSession)
-        cspScriptSrcHashes.push(`'${csp.hashAlgorithm}-${hash.digest('base64')}'`)
-      }
-
-      // Call ssr:csp hook
-      await this.context.nuxt.callHook('vue-renderer:ssr:csp', cspScriptSrcHashes)
+    // Skip the rest if SSR resources are not available
+    if (this.context.resources.ssrTemplate && this.context.resources.serverManifest) {
+      // Create bundle renderer for SSR
+      this.renderer.ssr = new SSRRenderer(this.context)
 
-      // Add csp meta tags
-      if (csp.addMeta) {
-        HEAD += `<meta http-equiv="Content-Security-Policy" content="script-src ${cspScriptSrcHashes.join()}">`
+      if (this.context.options.modern !== false) {
```

Comment:
Nit: do we have to check for `!== false` here or would a truth check be enough?

Reply:
!== false is more accurate here since modern can only be false, client or server

Reply:
Makes sense :+1: